In [1]:
import pandas as pd
df = pd.read_csv('../data/bronze/Tima_CRM - Data.csv')
df

,STT,SoTienDKVayBanDau,TienGiaiNgan,SoTienConLai,application_date,TS_CREDIT_SCORE_V2,Số điện thoại khách hàng,FromDate,ID,LoanID,...,InterestPaymentType,LongestOverdue,CreditInfo,Name,Address,CheckTime,Brieft,NumberOfLoans,HasBadDebt,HasLatePayment
0,2,10000000.0,NaN,0,2023-02-30,403,977966899,2016-06-24 00:00:00,40,16104,...,Lãi ngày,4,NaN,NaN,NaN,2019-03-17 11:22:15,notfound,-1,0,0
1,3,NaN,250000.0,0,2016-07-28,403,977966899,2016-07-28 00:00:00,226,20871,...,Lãi ngày,147,NaN,NaN,NaN,2019-03-17 11:22:15,notfound,-1,0,0
2,4,5000000.0,5000000.0,-50000,2016-06-29,531,397511119,2016-06-29 00:00:00,45,17049,...,Lãi ngày,3,"Khách hàng hiện không có quan hệ tại TCTD, khô...",DAO CONG QUYET,"P203-E9 PHUONG MAI,DONG DA,HA NOI",2019-03-17 08:29:24,0,0,0,0
3,5,10000000.0,10000000.0,0,2016-06-29,588,abc1234567,2016-06-29 00:00:00,48,17067,...,Lãi ngày,16,"Khách hàng hiện không có quan hệ tại TCTD, khô...",NGUYEN ANH VU,SO 21 NGACH 77/4 LE HONG PHONG-DIEN BIEN-BA DI...,2019-03-17 08:28:09,0,0,0,0
4,6,10000000.0,10000000.0,0,2016-08-05,675,966710180,2016-08-05 00:00:00,293,22390,...,Lãi ngày,8,"Khách hàng hiện không có quan hệ tại TCTD, khô...",ĐÀO NGỌC LINH,"-, TT CỤC KHÍ TƯỢNG THỦY VĂN, TRUNG LIỆT, Q.ĐỐ...",2019-03-17 14:14:22,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2378,3181,5000000.0,5000000.0,3474831,2018-12-24,577,353833091,2018-12-24 00:00:00,327685,56967,...,Dư nợ giảm dần cố định ngày đóng lãi,0,"Khách hàng hiện đang quan hệ tại 2 TCTD, không...",NGUYỄN HOÀI DUY,"40/8B,,TRẦN XUÂN SOẠN,TÂN THUẬN TÂY,QUẬN 7,TP ...",2019-03-17 11:57:59,2,2,0,0
2379,3182,3000000.0,3000000.0,3000000,2018-12-22,638,967295514,2018-12-22 00:00:00,329444,56835,...,Tất toán cuối kỳ,2,"Khách hàng hiện đang quan hệ tại 2 TCTD, không...",TRẦN HÙNG MẠNH,".,.,TỔ 32,PHƯỜNG NGUYỄN PHÚC,YÊN BÁI,TỈNH YÊN BÁI",2019-03-17 11:52:36,2,2,0,0
2380,3184,6000000.0,6000000.0,0,2018-12-27,605,385669091,2018-12-27 00:00:00,333862,57212,...,Tất toán cuối kỳ,0,"Khách hàng hiện đang quan hệ tại 1 TCTD, không...",LUU VAN HANG,"THANH NGHI, THANH XUAN, HA NOI",2019-03-17 11:43:43,1,1,0,0
2381,3187,7000000.0,6000000.0,0,2019-01-10,597,869625266,2019-01-10 00:00:00,348341,57938,...,Dư nợ giảm dần cố định ngày đóng lãi,0,"Khách hàng hiện đang quan hệ tại 2 TCTD, không...",NGUYỄN VĂN TRỌNG,THÔN LUẬT NGOẠI 2 - XÃ QUANG LỊCH,2019-03-17 10:47:41,2,2,0,0


In [2]:
df.shape

(2383, 49)

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2383 entries, 0 to 2382
Data columns (total 49 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   STT                          2383 non-null   str    
 1   SoTienDKVayBanDau            2382 non-null   float64
 2   TienGiaiNgan                 2382 non-null   float64
 3   SoTienConLai                 2383 non-null   int64  
 4   application_date             2383 non-null   str    
 5   TS_CREDIT_SCORE_V2           2383 non-null   int64  
 6   Số điện thoại khách hàng     2383 non-null   str    
 7   FromDate                     2383 non-null   str    
 8   ID                           2383 non-null   int64  
 9   LoanID                       2383 non-null   int64  
 10  Số tiền đăng ký vay ban đầu  2383 non-null   int64  
 11  Tiền giải ngân               2383 non-null   int64  
 12  Tiền gốc còn lại             2383 non-null   int64  
 13  ToDate                       

**Comments:**

- The data has 2383 rows and 49 columns.

- Some column data types need correction:

+, application_date: object --> date

+, Remaining amount: int64 --> float64

+, From Date: object --> datetime64

+, To Date: object --> datetime64

+, Gender: object --> Binary(0 - Female/1 - Male)

+, Birthday: object --> datetime64

+, Salary: object --> float64

+, CheckTime: object --> datetime64

+, HasBadDebt: int64 --> Binary (0 - No/1 - Yes)

+, HasLatePayment: int64 --> Binary (0 - No/1 - Yes)

In [4]:
#Reformat data
#Synchronize monetary data (convert to the same currency, same float)
import pandas as pd
import numpy as np
money_cols = [
    "SoTienDKVayBanDau", "TienGiaiNgan", "SoTienConLai", "Salary", "Số tiền đăng ký vay ban đầu", "Tiền giải ngân", "Tiền gốc còn lại"
]
money_cols = [c for c in money_cols if c in df.columns]
#Convert USD to VND, keep VND unchanged
def convert_to_vnd(x, usd_to_vnd=26000): # USD -> VND exchange rate
    if pd.isna(x):
        return np.nan
    if isinstance(x, (int, float)):
        return float(x)
    s = str(x).lower().strip()
    if "$" in s or "usd" in s:
        # Take the numerical value and multiply it by the USD exchange rate.
        s = s.replace("$", "").replace("usd", "").strip()
        try:
            return float(s) * usd_to_vnd
        except:
            return np.nan
    # If there is no $, consider it VND
    try:
        return float(s)
    except:
        return np.nan

for col in money_cols:
    df[col] = df[col].apply(convert_to_vnd).astype(float)

df[money_cols]

,SoTienDKVayBanDau,TienGiaiNgan,SoTienConLai,Salary,Số tiền đăng ký vay ban đầu,Tiền giải ngân,Tiền gốc còn lại
0,10000000.0,NaN,0.0,NaN,10000000.0,10000000.0,10000000.0
1,NaN,250000.0,0.0,7700000.0,10000000.0,10000000.0,0.0
2,5000000.0,5000000.0,-50000.0,26000000.0,5000000.0,5000000.0,5000000.0
3,10000000.0,10000000.0,0.0,NaN,10000000.0,10000000.0,10000000.0
4,10000000.0,10000000.0,0.0,12000000.0,10000000.0,10000000.0,0.0
...,...,...,...,...,...,...,...
2378,5000000.0,5000000.0,3474831.0,7500000.0,5000000.0,5000000.0,1814424.0
2379,3000000.0,3000000.0,3000000.0,NaN,3000000.0,3000000.0,3000000.0
2380,6000000.0,6000000.0,0.0,8000000.0,6000000.0,6000000.0,6000000.0
2381,7000000.0,6000000.0,0.0,12000000.0,7000000.0,6000000.0,6000000.0


In [5]:
#Change the data format of date columns to datetime
def parse_ymd_date(series):
    return pd.to_datetime(
        series.astype("string").str.slice(0, 10),
        errors="coerce",
        format="%Y-%m-%d"
    )

date_cols = ["application_date", "FromDate", "ToDate", "Birthday"]
for col in date_cols:
    df[col] = parse_ymd_date(df[col])

df["CheckTime"] = pd.to_datetime(
    df["CheckTime"],
    errors="coerce",
    format="%Y-%m-%d %H:%M:%S"
)

df[date_cols + ["CheckTime"]]


,application_date,FromDate,ToDate,Birthday,CheckTime
0,NaT,2016-06-24,2016-07-23,1987-12-16,2019-03-17 11:22:15
1,2016-07-28,2016-07-28,2017-03-24,1987-12-16,2019-03-17 11:22:15
2,2016-06-29,2016-06-29,2016-12-10,2025-01-01,2019-03-17 08:29:24
3,2016-06-29,2016-06-29,2016-10-26,1989-10-04,2019-03-17 08:28:09
4,2016-08-05,2016-08-05,2016-12-02,1982-01-31,2019-03-17 14:14:22
...,...,...,...,...,...
2378,2018-12-24,2018-12-24,2019-03-23,1996-12-09,2019-03-17 11:57:59
2379,2018-12-22,2018-12-22,2019-02-21,NaT,2019-03-17 11:52:36
2380,2018-12-27,2018-12-27,2019-03-26,1991-11-09,2019-03-17 11:43:43
2381,2019-01-10,2019-01-10,2019-04-09,1985-12-02,2019-03-17 10:47:41


In [6]:
df[["FullName", "Gender"]]

,FullName,Gender
0,Phạm Sơn Tùng,0
1,Phạm Sơn Tùng,Male
2,Đào Công Quyết,0
3,12345,0
4,Đào Ngọc Linh,1
...,...,...
2378,Nguyễn Hoài Duy,0
2379,Trần Hùng Mạnh,NaN
2380,Lưu Văn Hạng,0
2381,Nguyễn Văn Trọng,0


In [7]:
#Reformat some columns to Binary
import pandas as pd
import numpy as np

#Gender column (Gender: Male (Nam) --> 0, Female (Nữ) --> 1)
if "Gender" in df.columns:
    df["Gender"] = (
        df["Gender"]
        .astype(str)
        .str.strip()
        .str.lower()
        .replace({
            "nam": 0, "male": 0, "0":0,
            "nữ": 1,"female": 1, "1": 1
        })
    )
    df["Gender"] = pd.to_numeric(df["Gender"], errors="coerce").astype("Int64")

#HasBadDebt (1 = Yes, 0 = No)
if "HasBadDebt" in df.columns:
    df["HasBadDebt"] = (
        df["HasBadDebt"]
        .replace({"Có": 1, "Yes": 1, "yes": 1, True: 1, "1":1,
                  "Không": 0, "không": 0, "No": 0, "no": 0, False: 0, "0": 0})
    )
    df["HasBadDebt"] = pd.to_numeric(df["HasBadDebt"], errors="coerce").fillna(0).astype(int)

#HasLatePayment (1 = Yes, 0 = No)
if "HasLatePayment" in df.columns:
    df["HasLatePayment"] = (
        df["HasLatePayment"]
        .replace({"Có": 1, "co": 1, "có": 1, "Yes": 1, "yes": 1, True: 1,
                  "Không": 0, "không": 0, "No": 0, "no": 0, False: 0})
    )
    df["HasLatePayment"] = pd.to_numeric(df["HasLatePayment"], errors="coerce").fillna(0).astype(int)

df[["Gender", "HasBadDebt", "HasLatePayment"]]

,Gender,HasBadDebt,HasLatePayment
0,0,0,0
1,0,0,0
2,0,0,0
3,0,0,0
4,1,0,0
...,...,...,...
2378,0,0,0
2379,<NA>,0,0
2380,0,0,0
2381,0,0,0


In [8]:
#In the NumberOfLoans column, if the value is less than 0, change it to 0
if "NumberOfLoans" in df.columns:
    df.loc[df["NumberOfLoans"] < 0, "NumberOfLoans"] = 0

In [9]:
df["NumberOfLoans"]

0       0
1       0
2       0
3       0
4       0
       ..
2378    2
2379    2
2380    1
2381    2
2382    0
Name: NumberOfLoans, Length: 2383, dtype: int64

**Missing Value**

In [10]:
missing = df.isnull().sum().sort_values(ascending=False)
missing_ratio = (missing / len(df) * 100)
missing_df = pd.DataFrame({'MissingCount': missing, 'Missing%': missing_ratio})
missing_df

,MissingCount,Missing%
ReceiveYourIncomeSalary,1943,81.535879
WardName,1716,72.010071
CityCompany,610,25.597986
DescriptionPositionJob,482,20.226605
FullNameFamily,273,11.456148
JobName,248,10.407050
Address,218,9.148133
Name,217,9.106169
CreditInfo,217,9.106169
NameCompany,177,7.427612


**How ​​to handle missing data (missing values):**

- Remove columns with missing values ​​> 50%, including: ReceiveYourIncomeSalary (81.54%), WardName (72.01%)

- Drop the columns "SoTienDKVayBanDau", "TienGiaiNgan", "SoTienConLai" and use the "SoTienDKVayBanDau", "Tiền giải ngân", "SoTienConLai" columns to recalculate the value of the "SoTienConLai" column.

In [11]:
#Drop 2 two columns with high missing value rates
drop_cols = [c for c in ["WardName", "ReceiveYourIncomeSalary"] if c in df.columns]
df = df.drop(columns=drop_cols)

print("Row :", df.shape[0])
print("Column :", df.shape[1])

Row : 2383
Column : 47


In [12]:
#Drop cột SoTienDKVayBanDau, TienGiaiNgan, SoTienConLai
drop_cols = [c for c in ["SoTienDKVayBanDau", "TienGiaiNgan", "SoTienConLai"] if c in df.columns]
df = df.drop(columns=drop_cols)

print("Row :", df.shape[0])
print("Column :", df.shape[1])

Row : 2383
Column : 44


In [13]:
#Fill Salary = Average Salary by Card Number
df["Salary"] = df["Salary"].fillna(
    df.groupby("CardNumber")["Salary"].transform("mean")
)

df['Salary']

0        7700000.0
1        7700000.0
2       26000000.0
3              NaN
4       12000000.0
           ...    
2378     7500000.0
2379           NaN
2380     8000000.0
2381    12000000.0
2382    14000000.0
Name: Salary, Length: 2383, dtype: float64

In [14]:
#If there are candidates with the same salary and from the same company, fill in the JobName for the remaining blank JobNames.
import pandas as pd
import numpy as np

if "NameCompany" in df.columns:
    df["NameCompany"] = (df["NameCompany"].astype(str)
                                         .str.strip()
                                         .replace({"": np.nan, "None": np.nan, "nan": np.nan}))

# Only fill in if JobName is missing and NameCompany & Salary are complete.
mask = df["JobName"].isna() & df["NameCompany"].notna() & df["Salary"].notna()

def safe_mode(s: pd.Series):
    m = s.mode(dropna=True)
    return m.iloc[0] if len(m) else np.nan

# Calculate mode of JobName for (NameCompany, Salary)
group_mode = (
    df.groupby(["NameCompany", "Salary"])["JobName"]
      .transform(safe_mode)
)

before_na = df["JobName"].isna().sum()
df.loc[mask, "JobName"] = df.loc[mask, "JobName"].fillna(group_mode[mask])
after_na = df["JobName"].isna().sum()

In [15]:
# Fill Salary by mode of JobName
mode_by_job = df.groupby("JobName")["Salary"].agg(safe_mode)

df["Salary"] = df.apply(
        lambda row: mode_by_job[row["JobName"]] if pd.isna(row["Salary"]) and row["JobName"] in mode_by_job else row["Salary"],
        axis=1
    )

In [16]:
for c in df.select_dtypes(include=["object", "category"]).columns:
    df[c] = df[c].fillna("Unknown")

/var/folders/js/_vf645q54cz6vll84z7fp2gw0000gn/T/ipykernel_88063/1633594851.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for c in df.select_dtypes(include=["object", "category"]).columns:


In [17]:
df[['JobName', 'Salary']]

,JobName,Salary
0,Unknown,7700000.0
1,Unknown,7700000.0
2,Nhân viên chính thức,26000000.0
3,Nhân viên chính thức,8000000.0
4,Unknown,12000000.0
...,...,...
2378,Công nhân,7500000.0
2379,Unknown,NaN
2380,Kinh doanh,8000000.0
2381,Tự doanh không ĐKKD,12000000.0


In [18]:
#For any remaining salary blanks or zeros, fill in the median.
df['Salary'] = df['Salary'].replace(0, np.nan).fillna(df['Salary'].median())

In [19]:
df[['JobName', 'Salary']]

,JobName,Salary
0,Unknown,7700000.0
1,Unknown,7700000.0
2,Nhân viên chính thức,26000000.0
3,Nhân viên chính thức,8000000.0
4,Unknown,12000000.0
...,...,...
2378,Công nhân,7500000.0
2379,Unknown,8000000.0
2380,Kinh doanh,8000000.0
2381,Tự doanh không ĐKKD,12000000.0


In [20]:
#Change JobName 'Unknown' to 'Other'
df['JobName'] = df['JobName'].replace('Unknown', 'Khác')

In [21]:
#Change Residence Type 'Unknown' to 'Other'
df['Hình thức cư trú'] = df['Hình thức cư trú'].replace('Unknown', 'Khác')

In [22]:
#Missing value
missing = df.isnull().sum().sort_values(ascending=False)
missing_ratio = (missing / len(df) * 100)
missing_df = pd.DataFrame({'MissingCount': missing, 'Missing%': missing_ratio})
missing_df

,MissingCount,Missing%
CheckTime,147,6.168695
Birthday,4,0.167856
Gender,4,0.167856
application_date,1,0.041964
FullNameFamily,0,0.000000
NameCompany,0,0.000000
AddressCompany,0,0.000000
CityCompany,0,0.000000
DistrictNameCompany,0,0.000000
Salary,0,0.000000


**Duplicate**



In [23]:
dup_count = df.duplicated().sum()
print(f"Duplicate rows: {dup_count}")

Duplicate rows: 0


In [24]:
loanid_counts = df["LoanID"].value_counts()
dup_loanids = loanid_counts[loanid_counts > 1]
print(f"Duplicate LoanIDs: {len(dup_loanids)}")

Duplicate LoanIDs: 0


--> No two loan lines are alike, and no two loans are the same.

**Export to new dataset**

In [25]:
output_path = '../data/silver/tima_cleaned_data_v1.csv'
df.to_csv(output_path, index=False, encoding='utf-8-sig')